ouroboros-consensus-test/test-consensus/Test/Consensus/Mempool/StateMachine.hs

```diff
@@ -0,0 +1,810 @@
+     , ValidateEnvelope blk
+     )
+  => LedgerConfig blk
+  -> Model blk r
+  -> Command blk r
+  -> Response blk r
+  -> Model blk r
+transitions cfg model cmd resp = case (cmd, resp) of
+  (Action (NewTxs txs), Void) ->
+    if reachable
+    then
+      let nextTicket = succ $ modelLastTicket model
+          (res, tk, _, st) = foldTxs cfg nextTicket (currentRemainingCap model)
+                              mtip
+                            $ txs
+          modelTxs'     = mtxs ++
+                        zip [ txForgetValidated vtx
+                            | MempoolTxAdded vtx <- fst res
+                            ]
+                            [nextTicket..]
+      in
+        model {
+           modelMempoolTip = (st, reachable)
+         , modelTxs        = modelTxs'
+         , modelLastTicket = if tk == zeroTicketNo
+                             then zeroTicketNo
+                             else pred tk
+         , modelCapacity   = MempoolCapacityBytes 400
+         , modelMempoolWasFilled = modelMempoolWasFilled || (not . null . snd $ res)
+         }
+    else
+      let model' = transitions cfg model (Action SyncLedger) Void
+      in transitions cfg model' cmd Void
+
+  (Event (ChangeLedger l' b'), Void) ->
+        model { modelMempoolTip  = (mtip, reachable && not b')
+              , modelLedgerDBTip = l'
+              , modelSeenStates  = Set.insert (castPoint $ getTip l') seenStates
+              }
+  (Action GetSnapshot, GotSnapshot{}) -> model
+  (Action SyncLedger, Void) ->
+    if mtip == tick cfg dbtip
+    then model
+    else
+      let modelTxs'       = fmap fst mtxs
+          (res, _, _, st) = foldTxs cfg zeroTicketNo modelCapacity
+                              (tick cfg dbtip)
+                              modelTxs'
+          filteredTxSeq   =
+            [ (tx, tk) | (tx, tk) <- mtxs
+                          , tx
+                            `elem` [ txForgetValidated vtx
+                                   | MempoolTxAdded vtx <- fst res
+                                   ]
+                     ]
+      in
+        model {
+          modelMempoolTip = (st, True)
+        , modelTxs = filteredTxSeq
+        , modelCapacity = MempoolCapacityBytes 400
+        }
+
+  _ -> error $ "mismatched command "
+             <> show cmd
+             <> " and response "
+             <> show resp
+
+  where
+    Model {
+        modelMempoolTip  = (mtip, reachable)
+      , modelLedgerDBTip = dbtip
+      , modelTxs         = mtxs
+      , modelCapacity
+      , modelSeenStates  = seenStates
+      , modelMempoolWasFilled
+      } = model
+
+{-------------------------------------------------------------------------------
+  Ledger helper functions
+-------------------------------------------------------------------------------}
+
+-- | Apply a list of transactions short-circuiting if the mempool gets full.
+-- Emulates almost exactly the behavior of 'implTryAddTxs'.
```

Comment:
```suggestion
-- | Emulates almost exactly the behaviour of 'implTryAddTxs'.
```

I think we use UK eng. Eg `db-analyser`

Reply:
Perhaps we should explain the "almost" part. Ie which parts are not emulated and why.